codice/ddf
# DDF-04841 Add DDF support for phonetic matching

#### What does this PR do?
Adds DDF support for phonetic matching.  The capability is off by default but can be enabled through the admin console similar to the spellcheck capability.  Once enabled, the phonetic matching (termed Similar Word Matching in the UI) can be turned on or off per query and can be set as a user default search preference.
#### Who is reviewing it? 
@cantstoptheunk 
@andrewkfiedler 
@GabrielFabian 
@pklinef 
@bellcc 
@rzwiefel 
#### Select relevant component teams: 

@codice/solr 
@codice/ui 

#### Ask 2 committers to review/merge the PR and tag them here.

@andrewkfiedler
@bdeining
@jlcsmith

#### How should this be tested?
1. Log into the admin console and enable the `Enable Similar Word Matching` in the `Catalog Search UI`
2. Ingest a text based document with a set of words to be used for phonetic matching (e.g. `color`).
3. Search for one of the words in question with the same spelling and the similar matching turned off and verify that it can be found successfully
4. Change the spelling of the search term to another phonetic spelling (e.g. `colour`), re-run the search and verify that the ingested document is no longer returned
5. Enable the `Similar Word Matching` toggle on the query in the advanced query view, re-run the query, and verify that the ingested document is now returned again
6. Repeat for any other search terms for testing

Important note: Whether or not one word is a phonetic match to another one is dependent upon the phonetic algorithm being used.  Not all phonetically relevant matches will get returned in all cases and in some other cases false positives will be returned.  Just because one word doesn't phonetically match and return a result doesn't mean that it is not working correctly.  Most obvious phonetic matches will be returned as expected.
#### Any background context you want to provide?
#### What are the relevant tickets?

For GH Issues:
Fixes: #4841 ____

#### Screenshots

![image](https://user-images.githubusercontent.com/29106944/58151021-a6779600-7c25-11e9-895e-c4d8f381bda5.png)

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
